bench/meili/mapmatch.cc

```diff
@@ -0,0 +1,117 @@
+#include <iostream>
+#include <sstream>
+
+#include <benchmark/benchmark.h>
+#include <boost/property_tree/ptree.hpp>
+#include <rapidjson/rapidjson.h>
+
+#include "meili/map_matcher_factory.h"
+#include "meili/measurement.h"
+#include "sif/costconstants.h"
+#include "sif/costfactory.h"
+#include "tyr/actor.h"
+
+using namespace valhalla::midgard;
+using namespace valhalla::meili;
+using namespace valhalla::sif;
+
+namespace {
+
+#if !defined(VALHALLA_SOURCE_DIR)
+#define VALHALLA_SOURCE_DIR
+#endif
+
+constexpr float kGpsAccuracy = 4.07;
+constexpr float kSearchRadius = 50;
+
+// Inline benchmarks for OfflineMatch
+
+class OfflineMapmatchFixture : public benchmark::Fixture {
+public:
+  void SetUp(const ::benchmark::State& state) {
+    InitEngineConfig();
+    InitMapMatcher();
+  }
+
+  void TearDown(const ::benchmark::State& state) {
+    matcher_factory_->ClearCache();
+    mapmatcher_->Clear();
+    matcher_factory_.reset();
+    mapmatcher_.reset();
+  }
+
+private:
+  void InitEngineConfig() {
+    rapidjson::read_json(VALHALLA_SOURCE_DIR "bench/meili/config.json", config_);
+    const rapidjson::Document doc;
+    ParseAutoCostOptions(doc, "/costing_options/auto", options_.add_costing_options());
+  }
+
+  void InitMapMatcher() {
+    matcher_factory_ = std::make_shared<MapMatcherFactory>(config_);
+    mapmatcher_.reset(matcher_factory_->Create(valhalla::Costing::auto_, options_));
+  }
+
+protected:
+  // Mapmatcher implementation
+  std::shared_ptr<MapMatcher> mapmatcher_;
+  std::shared_ptr<MapMatcherFactory> matcher_factory_;
+  // Mapmatching configuration
+  boost::property_tree::ptree config_;
+  valhalla::Options options_;
+};
+
+std::vector<Measurement> BuildMeasurements(float gps_accuracy, float search_radius) {
+  std::vector<Measurement> meas;
+  meas.emplace_back(PointLL(5.09806, 52.09110), gps_accuracy, search_radius);
+  meas.emplace_back(PointLL(5.09769, 52.09050), gps_accuracy, search_radius);
+  meas.emplace_back(PointLL(5.09679, 52.09098), gps_accuracy, search_radius);
+  return meas;
+}
+
+BENCHMARK_DEFINE_F(OfflineMapmatchFixture, BasicOfflineMatch)(benchmark::State& state) {
+  const auto& meas = BuildMeasurements(kGpsAccuracy, kSearchRadius);
+  for (auto _ : state) {
+    benchmark::DoNotOptimize(mapmatcher_->OfflineMatch(meas));
+  }
+}
+
+BENCHMARK_REGISTER_F(OfflineMapmatchFixture, BasicOfflineMatch);
+
+// Load fixture files, intended to mirror test cases defined in test/mapmatch.cc.
+
+std::string LoadFile(const std::string filename) {
+  std::stringstream ss;
+  std::string line;
+  std::ifstream input_file;
+  input_file.open(filename.c_str());
+  while (std::getline(input_file, line)) {
+    ss << line;
+  }
+  return ss.str();
+}
+
+const std::vector<std::string> kBenchmarkCases = {
```

Comment:
@kevinkreiser, I'm commenting on the approach here in case it's not obvious: we're passing stand-alone `trace_route` requests as fixtures (as opposed to inlining them in the code) and then running an end-to-end benchmark that includes matching, serialization, etc. For an explanation about the use of `benchmark::DoNotOptimize(..)` on the expression return value, I'll point you to: https://github.com/google/benchmark#preventing-optimization.